sdk/python/kfp/v2/compiler/compiler.py

```diff
@@ -0,0 +1,318 @@
+    pipeline_meta = _extract_pipeline_metadata(pipeline_func)
+    pipeline_meta.name = pipeline_name or pipeline_meta.name
+    pipeline_meta.description = pipeline_description or pipeline_meta.description
+    pipeline_name = sanitize_k8s_name(pipeline_meta.name)
+
+    # Need to first clear the default value of dsl.PipelineParams. Otherwise, it
+    # will be resolved immediately in place when being to each component.
+    default_param_values = {}
+    for param in params_list:
+      default_param_values[param.name] = param.value
+      param.value = None
+
+    # Currently only allow specifying pipeline params at one place.
+    if params_list and pipeline_meta.inputs:
+      raise ValueError(
+          'Either specify pipeline params in the pipeline function, or in '
+          '"params_list", but not both.')
+
+    args_list = []
+    signature = inspect.signature(pipeline_func)
+    for arg_name in signature.parameters:
+      arg_type = None
+      for pipeline_input in pipeline_meta.inputs or []:
+        if arg_name == pipeline_input.name:
+          arg_type = pipeline_input.type
+          break
+      args_list.append(
+          dsl.PipelineParam(
+              sanitize_k8s_name(arg_name, True), param_type=arg_type))
+
+    with dsl.Pipeline(pipeline_name) as dsl_pipeline:
+      pipeline_func(*args_list)
+
+    # Fill in the default values.
+    args_list_with_defaults = []
+    if pipeline_meta.inputs:
+      args_list_with_defaults = [
+          dsl.PipelineParam(
+              sanitize_k8s_name(input_spec.name, True),
+              value=input_spec.default) for input_spec in pipeline_meta.inputs
+      ]
+    elif params_list:
+      # Or, if args are provided by params_list, fill in pipeline_meta.
+      for param in params_list:
+        param.value = default_param_values[param.name]
+
+      args_list_with_defaults = params_list
+      pipeline_meta.inputs = [
+          structures.InputSpec(
+              name=param.name, type=param.param_type, default=param.value)
+          for param in params_list
+      ]
+
+    pipeline_spec = self._create_pipeline_spec(
+        args_list_with_defaults,
+        dsl_pipeline,
+    )
+
+    return pipeline_spec
+
+  def compile(self,
+              pipeline_func: Callable[..., Any],
+              package_path: str,
+              type_check: bool = True) -> None:
+    """Compile the given pipeline function into workflow yaml.
+
+    Args:
+      pipeline_func: Pipeline function with @dsl.pipeline decorator.
+      package_path: The output pipeline spec tar.gz file path. for example,
+        "~/a.tar.gz"
+      type_check: Whether to enable the type check or not, default: True.
+    """
+    type_check_old_value = kfp.TYPE_CHECK
+    try:
+      kfp.TYPE_CHECK = type_check
+      self._create_and_write_pipeline_spec(
+          pipeline_func=pipeline_func, package_path=package_path)
+    finally:
+      kfp.TYPE_CHECK = type_check_old_value
+
+  @staticmethod
```

Comment:
Prefer to avoid staticmethod: https://google.github.io/styleguide/pyguide.html#217-function-and-method-decorators